File: RocaloidDevelopSuit/CVSEditor/CEditor/SynthesisEngine/Scheduler.vb
'    把所有合成器、缓冲器、效果器和CVS读写汇聚在一起的调度器。
'    The final class which puts all of the Synthesizers, Buffers, Effectors & CVS together.
'    Copyright (C) 2013 Sleepwalking
'
'    This program is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    This program is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <http://www.gnu.org/licenses/>.
Imports CybervoiceEngine
#Const DebugLevel = 1
Public Class Scheduler
	Public Shared CVS_ As CVSCommon.CVS
	Public Shared NowSynthesizing As Integer
	Public Shared StartSynthesize As Integer
	Private Shared Synth1 As SpeechSynthesizer
	'May enable Synth2 to achieve multi-thread in future.
	'Public Shared Synth2 As SpeechSynthesizer
	Private Shared Wave1 As WaveBuffer
	Private Shared Wave2 As WaveBuffer
	Private Shared SSSendBack As SpeechSynthesizerSendBack
	
	Public Shared Sub Init()
		Synth1 = New SpeechSynthesizer()
		'Synth2 = New SpeechSynthesizer()
		Wave1 = New WaveBuffer(30.0)
		Wave2 = New WaveBuffer(30.0)
		NowSynthesizing = 0
		StartSynthesize = 0
	End Sub
	Public Shared Sub OpenCVS(ByVal File As String)
		CVSCommon.Reader_Open(File)
		CVSCommon.CVS_Read(CVS_)
		CVSCommon.Reader_Close()
		#If DebugLevel > 0 Then
			CreateLog("Scheduler:	" & File & " loaded. " & CVS_.SegmentListQ + 1 & " segments in all.")
		#End If
	End Sub
	Public Shared Sub SetMemoryOutput(ByVal Buffer() As Byte)
		MixerWriterEffector.SetMemoryOutput(Buffer)
	End Sub
	Public Shared Sub RunSynthesizer()
		Dim S1 As Integer, S2 As Integer
		Dim i As Integer
		Dim T1 As Integer, T2 As Integer
		Dim Overlap As Integer
		T1 = SegmentSynthesize(Synth1, CVS_.SegmentList(NowSynthesizing), Wave1)
		S1 = CInt(CVS_.SegmentList(NowSynthesizing).StartTime * SampleRate) - StartSynthesize
		Overlap = 0
		MixerWriterEffector.WriteBlank(S1)
		For i = NowSynthesizing To CVS_.SegmentListQ - 1
			#If DebugLevel > 0 Then
				CreateLog("Scheduler:	Synthesizing segment " & i & ".")
			#End If
			S2 = CInt(CVS_.SegmentList(i + 1).StartTime * SampleRate) - StartSynthesize
			'Write to S2
			MixerWriterEffector.Write(Wave1, Overlap, S2 - S1)
			T2 = SegmentSynthesize(Synth1, CVS_.SegmentList(i + 1), Wave2)
			If S2 < S1 + T1 Then
				'Mix & Write to S1 + T1 if Overlapped
				MixerWriterEffector.MixAndWrite(Wave1, Wave2, S2 - S1, 0, S1 + T1 - S2)
				Overlap = S1 + T1 - S2
			Else
				Overlap = 0
			End If
			Swap(S1, S2)
			Swap(T1, T2)
			Swap(Wave1, Wave2)
		Next
		MixerWriterEffector.Write(Wave1, Overlap, T1 - Overlap)
		#If DebugLevel > 0 Then
			CreateLog("Scheduler:	Synthesis finished.")
		#End If
	End Sub
	Public Shared Function SegmentSynthesize(ByVal Synth As SpeechSynthesizer, _
										 ByVal _Segment As CVSCommon.Segment, _
										 ByVal Wave As WaveBuffer) As Integer
		Synth.SetSegment(_Segment)
		Wave.Pointer = 0
		Dim Frame As FrameBuffer
		Dim i As Integer = CInt(_Segment.StartTime * SampleRate)
		Dim j As Integer = CInt(CVSCommon.GetSegmentTime(_Segment) * SampleRate) + i
		While i < j
			Frame = Synth.Synthesize(i / SampleRate)
			Wave.Write(Frame)
			i += Frame.Length + 1
		End While
		SSSendBack = GlobalSendBack
		Effector.Render(Wave, _Segment, SSSendBack)
		Return Wave.Pointer
	End Function
End Class

File: RocaloidEngine/src/LibCVE/Scheduler.cc
#include "Scheduler.h"
#include "LibCyberBase/WaveBuffer.h"
#include "LibCVSCommon/CVSReader.h"
#include "SPKit/misc/converter.h"
#include "Effector/Effector.h"
#include "Effector/MixerWriter.h"

#define DebugLevel 1

#if(DebugLevel > 0)
	#include "SPKit/io/terminal.h"
#endif

using namespace converter;
namespace Scheduler
{
	CVSCommon::CVS* CVSData;
	SpeechSynthesizer* Synthesizer;
	WaveBuffer* Wave1;
	WaveBuffer* Wave2;
	SpeechSynthesizerSendBack SSSendBack;

	void Init()
	{
		Synthesizer = new SpeechSynthesizer();
		Wave1 = new WaveBuffer(30 * SampleRate);
		Wave2 = new WaveBuffer(30 * SampleRate);
		CVSData = new CVS;
	}
	void Exit()
	{
		delete Synthesizer;
		delete Wave1;
		delete Wave2;
		delete CVSData;
	}
	void CloseAll()
	{
		//Not ensured yet.
	}
	void OpenCVS(string FileName)
	{
		CVSReader TempReader;
		TempReader.Open(FileName);
		TempReader.Read(*CVSData);
		TempReader.Close();
		#if(DebugLevel > 1)
			wLine(CStr("Scheduler:  ") + FileName + CStr(" loaded. ") + 
		 	      CStr(CVSData -> SegmentListQ + 1) + CStr(" segments in all."));
		#endif
	}
	void SetFileOutput(string FileName)
	{
		MixerWriterEffector::SetFileOutput(FileName);
	}
	void RunSynthesizer()
	{
		int S1, S2;
		int T1, T2;
		int Overlap;
		int i;

		T1 = SegmentSynthesize(*Synthesizer, CVSData -> SegmentList[0], *Wave1);
		S1 = CInt(CVSData -> SegmentList[0].StartTime * SampleRate);
		Overlap = 0;
		MixerWriterEffector::WriteBlank(S1);
		//CVSData -> SegmentListQ
		for(i = 0;i < CVSData -> SegmentListQ;i ++)
		{
			#if(DebugLevel > 0)
				wLine(CStr("Scheduler: Synthesizing segment ") + CStr(i));
			#endif
			S2 = CInt(CVSData -> SegmentList[i + 1].StartTime * SampleRate);

			//Write to S2
			MixerWriterEffector::Write(*Wave1, Overlap, S2 - S1);
			T2 = SegmentSynthesize(*Synthesizer, CVSData -> SegmentList[i + 1], *Wave2);
			if(S2 < S1 + T1)
			{
				//Mix & Write to S1 + T1 if Overlapped
				MixerWriterEffector::MixAndWrite(*Wave1, *Wave2, S2 - S1, 0, S1 + T1 - S2);
				Overlap = S1 + T1 - S2;
			}else
				Overlap = 0;
			ObjSwap(S1, S2);
			ObjSwap(T1, T2);
			PtrSwap(Wave1, Wave2);
		}
		MixerWriterEffector::Write(*Wave1, Overlap, T1 - Overlap);
		MixerWriterEffector::Finish();
		#if(DebugLevel > 0)
			wLine(CStr("Scheduler: Synthesis finished."));
		#endif
	}
	int SegmentSynthesize(SpeechSynthesizer& Synth,
	                      CVSCommon::Segment& _Segment,
	                      WaveBuffer& Wave)
	{
		Synth.SetSegment(_Segment);
		Wave.Pointer = 0;
		FrameBuffer Frame;
		int i = CInt(_Segment.StartTime * SampleRate);
		int j = CInt(CVSCommon::GetSegmentTime(_Segment) * SampleRate) + i;
		while(i < j)
		{
			Synth.Synthesize(CDbl(i) / SampleRate, Frame);
			Wave.Write(Frame);
			i += Frame.Ubound + 1;
		}
		//SSSendBack = GlobalSendBack; Needn't since SSSendBack is in a namespace.
		Effector::Render(Wave, _Segment, SSSendBack);
		return Wave.Pointer;
	}
}